File: ambeth/Ambeth.Cache/ambeth/cache/interceptor/AbstractRootCacheAwareInterceptor.cs
using De.Osthus.Ambeth.Ioc;
using De.Osthus.Ambeth.Ioc.Annotation;
using De.Osthus.Ambeth.Ioc.Threadlocal;
using De.Osthus.Ambeth.Log;
using De.Osthus.Ambeth.Proxy;
using De.Osthus.Ambeth.Service;
using De.Osthus.Ambeth.Util;
using System.Reflection;
using System.Threading;

namespace De.Osthus.Ambeth.Cache.Interceptor
{
    public abstract class AbstractRootCacheAwareInterceptor : IThreadLocalCleanupBean
    {
        protected static readonly MethodInfo clearMethod;

        static AbstractRootCacheAwareInterceptor()
        {
            clearMethod = ReflectUtil.GetDeclaredMethod(false, typeof(IWritableCache), typeof(void), "Clear");
        }

        [LogInstance]
        public ILogger Log { private get; set; }

        [Autowired(Optional = true)]
        public IOfflineListenerExtendable OfflineListenerExtendable { protected get; set; }

        [Autowired]
        public IServiceContext ServiceContext { protected get; set; }

        [Autowired]
        public ICacheRetriever StoredCacheRetriever { protected get; set; }

        protected IRootCache AcquireRootCache(bool privileged, ThreadLocal<RootCache> currentRootCacheTL)
        {
            return AcquireRootCache(privileged, currentRootCacheTL, StoredCacheRetriever, null, null);
        }

        protected IRootCache AcquireRootCache(bool privileged, ThreadLocal<RootCache> currentRootCacheTL, ICacheRetriever cacheRetriever, Lock readLock, Lock writeLock)
        {
            IBeanRuntime<RootCache> rootCacheBR = ServiceContext.RegisterAnonymousBean<RootCache>().PropertyValue("CacheRetriever", cacheRetriever);
            if (readLock != null)
            {
                rootCacheBR.PropertyValue("ReadLock", readLock);
            }
            if (writeLock != null)
            {
                rootCacheBR.PropertyValue("WriteLock", writeLock);
            }
            RootCache rootCache = PostProcessRootCacheConfiguration(rootCacheBR).PropertyValue("Privileged", privileged).Finish();

            if (OfflineListenerExtendable != null)
            {
                OfflineListenerExtendable.AddOfflineListener(rootCache);
            }
            currentRootCacheTL.Value = rootCache;
            return rootCache;
        }

        protected IBeanRuntime<RootCache> PostProcessRootCacheConfiguration(IBeanRuntime<RootCache> rootCacheBR)
        {
            // Do not inject EventQueue because caches without foreign interest will never receive async DCEs
            return rootCacheBR.IgnoreProperties("EventQueue").PropertyValue("WeakEntries", false);
        }

        protected void DisposeCurrentRootCache(ThreadLocal<RootCache> currentTL)
        {
            RootCache rootCache = currentTL.Value;
            currentTL.Value = null;
            if (rootCache == null)
            {
                return;
            }
            if (OfflineListenerExtendable != null)
            {
                OfflineListenerExtendable.RemoveOfflineListener(rootCache);
            }
            // Cut reference to persistence layer
            rootCache.Dispose();
        }

        public abstract void CleanupThreadLocal();
    }
}

File: jambeth/jambeth-cache/src/main/java/de/osthus/ambeth/cache/interceptor/AbstractRootCacheAwareInterceptor.java
package de.osthus.ambeth.cache.interceptor;

import java.lang.reflect.Method;

import de.osthus.ambeth.cache.IRootCache;
import de.osthus.ambeth.cache.IWritableCache;
import de.osthus.ambeth.cache.RootCache;
import de.osthus.ambeth.exception.RuntimeExceptionUtil;
import de.osthus.ambeth.ioc.IBeanRuntime;
import de.osthus.ambeth.ioc.IServiceContext;
import de.osthus.ambeth.ioc.annotation.Autowired;
import de.osthus.ambeth.ioc.threadlocal.IThreadLocalCleanupBean;
import de.osthus.ambeth.log.ILogger;
import de.osthus.ambeth.log.LogInstance;
import de.osthus.ambeth.proxy.AbstractSimpleInterceptor;
import de.osthus.ambeth.service.ICacheRetriever;
import de.osthus.ambeth.service.IOfflineListenerExtendable;
import de.osthus.ambeth.util.Lock;

public abstract class AbstractRootCacheAwareInterceptor extends AbstractSimpleInterceptor implements IThreadLocalCleanupBean
{
	protected static final Method clearMethod;

	static
	{
		try
		{
			clearMethod = IWritableCache.class.getMethod("clear");
		}
		catch (SecurityException e)
		{
			throw RuntimeExceptionUtil.mask(e);
		}
		catch (NoSuchMethodException e)
		{
			throw RuntimeExceptionUtil.mask(e);
		}
	}

	@SuppressWarnings("unused")
	@LogInstance
	private ILogger log;

	@Autowired(optional = true)
	protected IOfflineListenerExtendable offlineListenerExtendable;

	@Autowired
	protected IServiceContext serviceContext;

	@Autowired
	protected ICacheRetriever storedCacheRetriever;

	protected IRootCache acquireRootCache(boolean privileged, ThreadLocal<RootCache> currentRootCacheTL)
	{
		return acquireRootCache(privileged, currentRootCacheTL, storedCacheRetriever, null, null);
	}

	protected IRootCache acquireRootCache(boolean privileged, ThreadLocal<RootCache> currentRootCacheTL, ICacheRetriever cacheRetriever, Lock readLock,
			Lock writeLock)
	{
		IBeanRuntime<RootCache> rootCacheBR = serviceContext.registerBean(RootCache.class).propertyValue("CacheRetriever", cacheRetriever);
		if (readLock != null)
		{
			rootCacheBR.propertyValue("ReadLock", readLock);
		}
		if (writeLock != null)
		{
			rootCacheBR.propertyValue("WriteLock", writeLock);
		}
		RootCache rootCache = postProcessRootCacheConfiguration(rootCacheBR).propertyValue("Privileged", Boolean.valueOf(privileged)).finish();

		if (offlineListenerExtendable != null)
		{
			offlineListenerExtendable.addOfflineListener(rootCache);
		}
		currentRootCacheTL.set(rootCache);
		return rootCache;
	}

	protected IBeanRuntime<RootCache> postProcessRootCacheConfiguration(IBeanRuntime<RootCache> rootCacheBR)
	{
		// Do not inject EventQueue because caches without foreign interest will never receive async DCEs
		return rootCacheBR.ignoreProperties("EventQueue").propertyValue("WeakEntries", Boolean.FALSE);
	}

	protected void disposeCurrentRootCache(ThreadLocal<RootCache> currentTL)
	{
		RootCache rootCache = currentTL.get();
		currentTL.remove();
		if (rootCache == null)
		{
			return;
		}
		if (offlineListenerExtendable != null)
		{
			offlineListenerExtendable.removeOfflineListener(rootCache);
		}
		// Cut reference to persistence layer
		rootCache.dispose();
	}
}